### PR TITLE
Reduce library search path length in new build dir layout

### PR DESCRIPTION
**Pull request by ranger-ross**
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/cargo/pull/17191)*

### What does this PR try to resolve?

This PR addresses #17183 by only adding dylib crates to the library search path (`LD_LIBRARY_PATH` on linux, `PATH` on windows)

Previously, `deps` would be added to the search path. With the new build dir layout, I naively added all `out` dirs which blew up the search path hitting OS limits.

See https://github.com/rust-lang/cargo/issues/17183#issuecomment-4917589909 for more details.

### How to test and review this PR?

See the included test. I also verified that Zed now properly compiles with the new build-dir layout (on Linux)

### CODE REVIEW

File: tests/testsuite/build.rs [hunk 1]
```diff
@@ -6598,3 +6598,69 @@ fn should_not_include_build_script_out_dir_path_in_rustc_args() {
 "#]])
         .run();
 }
+
+#[cargo_test]
+fn should_only_include_dylibs_on_lib_search_path() {
```
Comment (weihanglo) [inline tests/testsuite/build.rs:6607]: What about a dep has `cargo::rustc-link-search` in its build.rs, or a transitive dep has that, but the dep itself is just an rlib.

(I feel like those has been covered already though)

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/cargo/17191/b36402f8f770a0ee97b17460e3b429268e4c2527..0b2d47020d284b59bfa0b8781acfe3e16ca73312)*

Comment (ranger-ross) [reply tests/testsuite/build.rs:6607]: I mentioned this in the issue, but I believe build script directives will be unaffected by this change. I didn't think to test this but I'd also like to think that this is an existing test. 

I'll take a look to double check later

Comment (ranger-ross) [reply tests/testsuite/build.rs:6607]: I think [`linker_search_path_preference`](https://github.com/rust-lang/cargo/blob/0a28f7930c7b559c37fc221347114f9c6434f2ae/tests/testsuite/build_script.rs#L6730) covers a good amount of scenarios.

Though, happy to add more tests to cover more edge cases

Comment (weihanglo) [reply tests/testsuite/build.rs:6607]: Thanks. Will review this closer later today. Need to get food to feed myself now 😬

**Comment by ranger-ross** [comment]
I noticed the test was failing on stable, due to the behavior depending on https://github.com/rust-lang/rust/pull/155439/changes/61f3e086acc1c187bb262ab43cac71f44018c397 which still has not landed on stable.
I made the test nightly only until that commit lands on stable rust, in which case we can remove nightly only status

**Comment by ranger-ross** [comment]
Marking this as ready for review, the tests passed in CI and the CI failure appears to just be a spurious 503 from crates.io

File: src/cargo/core/compiler/build_runner/mod.rs [hunk 2]
```diff
@@ -485,11 +485,13 @@ impl<'a, 'gctx> BuildRunner<'a, 'gctx> {
                 for (unit, _) in self.bcx.unit_graph.iter() {
                     let dep_dir = self.files().deps_dir(unit);
                     paths::create_dir_all(&dep_dir)?;
-                    self.compilation
-                        .deps_output
-                        .entry(kind)
-                        .or_default()
-                        .insert(dep_dir);
+                    if unit.target.is_dylib() {
```
Comment (weihanglo) [inline src/cargo/core/compiler/build_runner/mod.rs:491]: I feel like maybe at some point we should rename `deps_output` to `dylibs_path` or something, and update its doc comment.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/cargo/17191/0a28f7930c7b559c37fc221347114f9c6434f2ae..1411fd1885fd29de0e1998dbb97a88cb4e420a3f)*

File: tests/testsuite/build.rs [hunk 3]
```diff
@@ -6598,3 +6598,71 @@ fn should_not_include_build_script_out_dir_path_in_rustc_args() {
...
+            "#,
+        )
+        .file("baz/src/lib.rs", r#"pub fn baz_value() -> i32 { 200 }"#)
+        .build();
+
+    p.cargo("-Zbuild-dir-new-layout -v run")
+        .masquerade_as_nightly_cargo(&["new build-dir layout"])
+        .enable_mac_dsym()
+        .with_stdout_data(str![[r#"
+[..][ROOT]/foo/target/debug/build/bar/[HASH]/out[..][ROOT]/foo/target/debug/build/baz/[HASH]/out[..]
```
Comment (weihanglo) [inline tests/testsuite/build.rs]: I believe this is asserting rustc invocation `Running rustc…` right, could you have the prefix so we know this is it.

(I was confused for a minute because `fill_env` is also used in `cargo run` runner and this executes a `cargo run`).

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/cargo/17191/0a28f7930c7b559c37fc221347114f9c6434f2ae..1411fd1885fd29de0e1998dbb97a88cb4e420a3f)*

File: src/cargo/core/compiler/build_runner/mod.rs (hunk 2 above)
Comment (weihanglo) [inline src/cargo/core/compiler/build_runner/mod.rs:491]: This is probably an edge case but with `cdylib` the search path change, we should probably append if rdylib or cydlib?.

```rust
#[cargo_test]
fn cdylib_on_search_path() {
    let envvar = cargo_util::paths::dylib_path_envvar();
    let p = project()
        .file(
            "Cargo.toml",
            r#"
                [package]
                name = "foo"
                edition = "2021"

                [dependencies]
                bar = { path = "bar" }
            "#,
        )
        .file(
            "src/main.rs",
            &format!(
                r#"
                    fn main() {{
                        let search_path = std::env::var_os("{envvar}").unwrap();
                        let found = std::env::split_paths(&search_path)
                            .any(|dir| dir.join("{DLL_PREFIX}bar{DLL_SUFFIX}").exists());
                        println!("cdylib on search path: {{found}}");
                    }}
                "#
            ),
        )
        .file(
            "bar/Cargo.toml",
            r#"
                [package]
                name = "bar"
                edition = "2021"

                [lib]
                crate-type = ["lib", "cdylib"]
            "#,
        )
        .file("bar/src/lib.rs", r#"pub fn bar_value() -> i32 { 100 }"#)
        .build();

    p.cargo("run")
        .with_stdout_data(str![[r#"
cdylib on search path: true

"#]])
        .run();

    p.cargo("run -Zbuild-dir-new-layout")
        .masquerade_as_nightly_cargo(&["build-dir-new-layout"])
        .enable_mac_dsym()
        .with_stdout_data(str![[r#"
cdylib on search path: false

"#]])
        .run();
}
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/cargo/17191/0a28f7930c7b559c37fc221347114f9c6434f2ae..1411fd1885fd29de0e1998dbb97a88cb4e420a3f)*

Comment (ranger-ross) [reply src/cargo/core/compiler/build_runner/mod.rs:491]: Yeah, I also thought about that since its only really ever used for dylibs.
Though I think we should probably leave it as is until the new layout is stablized.

I added this comment to the checklist in https://github.com/rust-lang/cargo/issues/17182

File: src/cargo/core/compiler/build_runner/mod.rs [hunk 4]
```diff
@@ -485,11 +485,13 @@ impl<'a, 'gctx> BuildRunner<'a, 'gctx> {
                 for (unit, _) in self.bcx.unit_graph.iter() {
```
Comment (weihanglo) [inline src/cargo/core/compiler/build_runner/mod.rs:485]: Out of scope of the fix but maybe this?

```suggestion
                for (unit, _) in self.bcx.unit_graph.iter() {
                   if kind != unit.kind {
                       continue;
                   }
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/cargo/17191/0a28f7930c7b559c37fc221347114f9c6434f2ae..1411fd1885fd29de0e1998dbb97a88cb4e420a3f)*

File: tests/testsuite/build.rs (hunk 3 above)
Comment (ranger-ross) [reply tests/testsuite/build.rs]: actually its `stdout`, not `stderr`. So its asserting the output of `cargo run` which is:
```rs
 fn main() {{
    println!("{{}}", env!("{envvar}"));
}}
```

Wasn't sure if there was a better to assert env vars for commands.

I left off the `with_stdout_data` assertion since I felt it was not relavent to what the test was trying to assert, but I can add it if you think that would make the test easier to understand

Comment (ranger-ross) [reply tests/testsuite/build.rs]: I think I'll edit the test to be more like your proposed test in https://github.com/rust-lang/cargo/pull/17191#discussion_r3564742137 since I think it makes it clearer what is being asserted

File: tests/testsuite/build.rs [hunk 5]
```diff
@@ -6598,3 +6599,109 @@ fn should_not_include_build_script_out_dir_path_in_rustc_args() {
...
+        )
+        .file("baz/src/lib.rs", r#"pub fn baz_value() -> i32 { 200 }"#)
+        .file(
+            "qux/Cargo.toml",
+            r#"
+                [package]
+                name = "qux"
+                version = "0.1.0"
+                edition = "2021"
+                authors = []
```
Comment (weihanglo) [inline tests/testsuite/build.rs:6691]: You probably forgot to add 

```
[lib]
crate-type = ["cdylib"]
```

File: src/cargo/core/compiler/build_runner/mod.rs (hunk 2 above)
Comment (ranger-ross) [reply src/cargo/core/compiler/build_runner/mod.rs:491]: I like the idea of also running with the old layout as well so included that in the test.
Actually it appears that even in the old layout, cdylibs were not on the search path which is interesting.

File: src/cargo/core/compiler/build_runner/mod.rs (hunk 4 above)
Comment (ranger-ross) [reply src/cargo/core/compiler/build_runner/mod.rs:485]: good catch!

I made this change and included a test for this as well.
I think the easiest way to test this was to have a build script (which will always be a host kind) and cross compile to another target. Then verify that the dylibs for the cross compiled target are not on the build script search path.

File: src/cargo/core/compiler/build_runner/mod.rs
Comment (weihanglo) [inline src/cargo/core/compiler/build_runner/mod.rs:1]: BTW, I just got another idea: the current `deps_output` map collects every dylib output dir from every unit in the compilation session, which is not ideal. Some of the unit only need a subtree of it, and would be basically zero dylib path.

Anyway, I believe we don't need to put that much effort on this becuase dylibs are not that common as dependencies.

**Review by weihanglo — COMMENTED** [review]
Let me know when it is ready (pretty close I believe). Also, if we want to unblock https://github.com/rust-lang/rust/issues/159076 we should do beta backport, right?

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/cargo/17191/2ceefa0090080354b80cc2f5415039bdb0d2bf0b..50d9cbfe784f62a58e26c317c292fd661455eb81)*

File: tests/testsuite/build.rs (hunk 5 above)
Comment (ranger-ross) [reply tests/testsuite/build.rs:6691]: oops, I must have lost that while rebasing. Thanks. :sweat_smile:

File: src/cargo/core/compiler/build_runner/mod.rs
Comment (ranger-ross) [reply src/cargo/core/compiler/build_runner/mod.rs:1]: Hmmm yeah, good point. I probably added `unit_graph.iter()` to cast a wide net as possible.
I think we could construct lists per `Unit` instead of `CompileKind`.

Though, like you mentioned, dylibs usually don't make up a large percent of the crate graph so not sure if the effort for that optimization would be worth it

**Comment by ranger-ross** [comment]
> Let me know when it is ready (pretty close I believe). Also, if we want to unblock https://github.com/rust-lang/rust/issues/159076 we should do beta backport, right?

I think its ready if it looks good to you.
Yeah, I think this will need to be backported. 

I can raise a PR for that was well

**Review by weihanglo — APPROVED** [review]
Thanks!

Yeah would be nice to have a beta backport.



*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/cargo/17191/2ceefa0090080354b80cc2f5415039bdb0d2bf0b..1b4c3e8d3061e09882d4f18c2daa05882ecd2ca4)*
